rewrite-maven/src/main/java/org/openrewrite/maven/RemoveDuplicateDependencies.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2020 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.maven;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.internal.ListUtils;
+import org.openrewrite.xml.XPathMatcher;
+import org.openrewrite.xml.tree.Xml;
+
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.UnaryOperator;
+
+@Value
+@EqualsAndHashCode(callSuper = true)
+public class RemoveDuplicateDependencies extends Recipe {
+
+    @Override
+    public String getDisplayName() {
+        return "Remove duplicate Maven dependencies";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Removes duplicated dependencies in the <dependencies> section of the pom.xml.";
+    }
+
+    @Override
+    public Duration getEstimatedEffortPerOccurrence() {
+        return Duration.ofMinutes(2);
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new RemoveDuplicateDependenciesVisitor();
+    }
+
+    @Override
+    protected MavenIsoVisitor<ExecutionContext> getApplicableTest() {
+        return new MavenIsoVisitor<ExecutionContext>() {
+            @Override
+            public Xml.Tag visitTag(Xml.Tag tag, ExecutionContext context) {
+                if ("project".equals(tag.getName()) && getCursor().getParent().getValue() instanceof Xml.Document) {
+                    if (tag.getChild("dependencies").isPresent() || tag.getChild("dependencyManagement").isPresent()) {
+                        // dummy change to indicate that this recipe is applicable
+                        return tag.withId(UUID.randomUUID());
```

Comment:
I wasn't quite sure how this test should be implemented, so this is what I came up with. The corresponding test just uses `cycles(0)` to check that the recipe never gets executed.

Reply:
Your implementation is pretty close and provides an excellent opportunity to learn some key aspects of OpenRewrite and the AST.  

The AST data model is immutable by design, so simply adding a `SearchMarker` to any node is sufficient for detecting a change and allowing the recipe to run.  Here is how to add a search marker
```java
if (root.getChild("dependencies").isPresent() || root.getChild("dependencyManagement").isPresent()) {
    SearchResult.found(root);
}
```

Here is where the RecipeScheduler applies the `singleSourceApplibilityTest` to each source file checking for referential equality.
https://github.com/openrewrite/rewrite/blob/46d0578cfb597fe013f54de924bd17b5cec6c0bb/rewrite-core/src/main/java/org/openrewrite/RecipeScheduler.java#L206-L211

In most cases an ASTs `ID` is changed when its being copied into a new element, such as in the [ReplaceDuplicateStringLiterals](https://github.com/openrewrite/rewrite/blob/46d0578cfb597fe013f54de924bd17b5cec6c0bb/rewrite-java/src/main/java/org/openrewrite/java/cleanup/ReplaceDuplicateStringLiterals.java#L116) recipe. 

Notice that Lombok `@EqualsAndHashCode(callSuper = false, onlyExplicitlyIncluded = true)` annotations on every tree. 
https://github.com/openrewrite/rewrite/blob/46d0578cfb597fe013f54de924bd17b5cec6c0bb/rewrite-xml/src/main/java/org/openrewrite/xml/tree/Xml.java#L76-L82

Reply:
Thanks for the tip with `SearchResult#found()`. I suspected there would be a better way :-)